File: gateway/utils/harper_canonical_validation.py
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Dict, Iterable, List, Optional


PROMPT_TEMPLATE_PHRASES = [
    "Print EXCLUSIVELY " + "one file block",
    "Produce only " + "the single",
    "single " + "BEGIN_FILE",
    "No additional " + "files",
]

UNRESOLVED_PLACEHOLDERS = [
    "<Project Name>",
    "<X min",
    "<...>",
    "My Solution Name",
]

GENERIC_TEMPLATE_ONLY_VALUES = [
    "aws-eks",
    "my-project-key",
    "https://api.openai.com",
]

BMAD_IDEA_EXTRA_SECTIONS = [
    "## Deployment Portability Rule",
    "## Technology Constraints Profile Rule",
    "## Strategic Fit",
    "## /spec Handoff Readiness",
]


def _result(path: str, checks: List[str], diagnostic: str) -> Dict[str, Any]:
    return {
        "ok": not checks,
        "path": path,
        "failed_checks": checks,
        "diagnostic": diagnostic if checks else "",
        "error_code": "invalid_canonical_artifact" if checks else None,
    }


def _contains_case_insensitive(text: str, needle: str) -> bool:
    return needle.lower() in text.lower()


def _has_heading(text: str, heading: str) -> bool:
    pattern = rf"(?im)^\s*{re.escape(heading)}\s*$"
    return re.search(pattern, text) is not None


def _heading_index(text: str, heading: str) -> int:
    match = re.search(rf"(?im)^\s*{re.escape(heading)}\s*$", text)
    return match.start() if match else -1


def _has_any_heading(text: str, headings: Iterable[str]) -> bool:
    return any(_has_heading(text, heading) for heading in headings)


def _starts_with_heading(text: str, prefix: str) -> bool:
    first = str(text or "").lstrip().splitlines()[0] if str(text or "").strip() else ""
    return first.startswith(prefix)


def _common_markdown_failures(content: str, evidence_text: str = "") -> List[str]:
    text = str(content or "")
    failures: List[str] = []
    if "BEGIN_FILE" in text:
        failures.append("contains_BEGIN_FILE")
    if "END_FILE" in text:
        failures.append("contains_END_FILE")
    for phrase in PROMPT_TEMPLATE_PHRASES:
        if _contains_case_insensitive(text, phrase):
            failures.append(f"contains_prompt_template_phrase:{phrase}")
    for placeholder in UNRESOLVED_PLACEHOLDERS:
        if placeholder in text:
            failures.append(f"contains_unresolved_placeholder:{placeholder}")
    for value in GENERIC_TEMPLATE_ONLY_VALUES:
        if value in text and value not in str(evidence_text or ""):
            failures.append(f"contains_unevidenced_template_value:{value}")
    return failures


def _looks_like_fenced_yaml(text: str) -> bool:
    return re.search(r"(?im)^```ya?ml\s*$", text) is not None


def _section_after(text: str, heading: str) -> str:
    match = re.search(rf"(?im)^\s*{re.escape(heading)}\s*$", text)
    if not match:
        return ""
    rest = text[match.end():]
    next_heading = re.search(r"(?m)^\s*##\s+", rest)
    return rest[: next_heading.start()] if next_heading else rest


def validateIdeaMarkdown(
    content: str,
    *,
    path: str = "docs/harper/IDEA.md",
    evidence_text: str = "",
) -> Dict[str, Any]:
    text = str(content or "")
    stripped = text.lstrip()
    failures = _common_markdown_failures(text, evidence_text)

    if not (stripped.startswith("# IDEA — ") or stripped.startswith("# IDEA - ")):
        failures.append("missing_idea_h1")
    if stripped.startswith("```") or stripped.lower().startswith("tech_constraints:"):
        failures.append("starts_with_raw_yaml_or_fence")

    primary_headings = [
        "## Vision",
        "## Problem Statement",
        "## Target Users & Context",
        ("## Value & Outcomes", "## Value & Outcomes (with initial targets)"),
        ("## Out of Scope", "## Out of Scope (slice-1)"),
        ("## Technology Constraints", "## Technology Constraints (SPEC-ready)"),
        "## Risks & Assumptions",
    ]

    primary_indexes: List[int] = []
    for item in primary_headings:
        if isinstance(item, tuple):
            found = [idx for idx in (_heading_index(text, option) for option in item) if idx >= 0]
            if not found:
                failures.append("missing_heading:" + "|".join(item))
            else:
                primary_indexes.append(min(found))
        else:
            idx = _heading_index(text, item)
            if idx < 0:
                failures.append(f"missing_heading:{item}")
            else:
                primary_indexes.append(idx)

    if primary_indexes and primary_indexes != sorted(primary_indexes):
        failures.append("primary_idea_sections_out_of_order")

    tech_section = _section_after(text, "## Technology Constraints") or _section_after(text, "## Technology Constraints (SPEC-ready)")
    if not _looks_like_fenced_yaml(tech_section):
        failures.append("missing_fenced_yaml_after_technology_constraints")

    if not re.search(r"(?im)^\s*##\s+.*success metrics", text):
        failures.append("missing_success_metrics_section")

    core_opening_indexes = [
        idx
        for heading in ["## Vision", "## Problem Statement", "## Target Users & Context"]
        for idx in [_heading_index(text, heading)]
        if idx >= 0
    ]
    if core_opening_indexes:
        last_core_opening = max(core_opening_indexes)
        for heading in BMAD_IDEA_EXTRA_SECTIONS:
            idx = _heading_index(text, heading)
            if 0 <= idx < last_core_opening:
                failures.append(f"bmad_extra_section_before_primary:{heading}")

    return _result(path, failures, "IDEA.md failed canonical Harper structure validation.")


def validateSpecMarkdown(content: str, *, path: str = "docs/harper/SPEC.md") -> Dict[str, Any]:
    text = str(content or "")
    lower = text.lower()
    failures = _common_markdown_failures(text)
    if not _starts_with_heading(text, "# SPEC"):
        failures.append("missing_spec_h1")
    if not any(term in lower for term in ["problem", "scope", "objective", "requirement"]):
        failures.append("missing_problem_scope_or_requirements")
    if not any(term in lower for term in ["acceptance criteria", "testability", "test strategy"]):
        failures.append("missing_acceptance_or_testability")
    if not any(term in lower for term in ["constraint", "non-functional", "non functional"]):
        failures.append("missing_constraints_or_non_functional_requirements")
    if "spec_ux_appendix" in lower or "user journey" in lower and "functional requirement" not in lower:
        failures.append("looks_like_companion_only_ux_content")
    return _result(path, failures, "SPEC.md failed canonical Harper structure validation.")


def validatePlanMarkdown(content: str, *, path: str = "docs/harper/PLAN.md") -> Dict[str, Any]:
    text = str(content or "")
    lower = text.lower()
    failures = _common_markdown_failures(text)
    if not _starts_with_heading(text, "# PLAN"):
        failures.append("missing_plan_h1")
    if "REQ-" not in text:
        failures.append("missing_req_ids")
    if not any(term in lower for term in ["dependencies", "dependson", "ordering", "depends on"]):
        failures.append("missing_dependencies_or_ordering")
    if not any(term in lower for term in ["/kit", "kit readiness", "kit-readiness", "implementation readiness"]):
        failures.append("missing_kit_or_implementation_readiness")
    return _result(path, failures, "PLAN.md failed canonical Harper structure validation.")


def validatePlanJson(content: str, *, path: str = "docs/harper/plan.json") -> Dict[str, Any]:
    text = str(content or "")
    failures: List[str] = []
    if text.lstrip().startswith("#") or "```" in text:
        failures.append("looks_like_markdown")
    try:
        data = json.loads(text)
    except Exception:
        return _result(path, [*failures, "invalid_json"], "plan.json must be valid JSON.")

    if not isinstance(data, dict):
        failures.append("json_root_not_object")
        return _result(path, failures, "plan.json root must be an object.")

    reqs = data.get("reqs") or data.get("requirements") or data.get("items")
    if not isinstance(reqs, list) or not reqs:
        failures.append("missing_requirements_list")
    else:
        for index, req in enumerate(reqs):
            if not isinstance(req, dict):
                failures.append(f"req_{index}_not_object")
                continue
            for field in ["id", "title", "status", "acceptance"]:
                if not req.get(field):
                    failures.append(f"req_{index}_missing_{field}")
            if "dependsOn" not in req and "dependencies" not in req and "depends_on" not in req:
                failures.append(f"req_{index}_missing_dependencies")

    return _result(path, failures, "plan.json failed canonical Harper structure validation.")


def validateLaneGuideMarkdown(content: str, *, path: str = "docs/harper/lane-guides/<lane>.md") -> Dict[str, Any]:
    text = str(content or "")
    lower = text.lower()
    failures = _common_markdown_failures(text)
    if not text.lstrip().startswith("#"):
        failures.append("missing_heading")
    if not any(term in lower for term in ["lane purpose", "purpose", "scope"]):
        failures.append("missing_lane_purpose_or_scope")
    if not any(term in lower for term in ["expected files", "boundaries", "boundary"]):
        failures.append("missing_expected_files_or_boundaries")
    if not any(term in lower for term in ["test command", "validation command", "commands"]):
        failures.append("missing_test_or_validation_commands")
    if not any(term in lower for term in ["eval/gate", "eval expectations", "gate expectations"]):
        failures.append("missing_eval_gate_expectations")
    return _result(path, failures, "Lane guide failed canonical Harper structure validation.")


def validate_canonical_harper_artifact(path: str, content: str, *, evidence_text: str = "") -> Optional[Dict[str, Any]]:
    normalized = str(path or "").replace("\\", "/").lstrip("./").lstrip("/")
    bare_to_canonical = {
        "IDEA.md": "docs/harper/IDEA.md",
        "SPEC.md": "docs/harper/SPEC.md",
        "PLAN.md": "docs/harper/PLAN.md",
        "plan.json": "docs/harper/plan.json",
    }
    normalized = bare_to_canonical.get(normalized, normalized)
    if normalized == "docs/harper/IDEA.md":
        return validateIdeaMarkdown(content, path=normalized, evidence_text=evidence_text)
    if normalized == "docs/harper/SPEC.md":
        return validateSpecMarkdown(content, path=normalized)
    if normalized == "docs/harper/PLAN.md":
        return validatePlanMarkdown(content, path=normalized)
    if normalized == "docs/harper/plan.json":
        return validatePlanJson(content, path=normalized)
    if re.fullmatch(r"docs/harper/lane-guides/[^/]+\.md", normalized):
        return validateLaneGuideMarkdown(content, path=normalized)
    return None


def validate_canonical_harper_files(
    files: List[Dict[str, Any]],
    *,
    evidence_text: str = "",
) -> Dict[str, Any]:
    accepted: List[Dict[str, Any]] = []
    rejected: List[Dict[str, Any]] = []
    rejected_source_files: List[Dict[str, Any]] = []
    for item in files or []:
        path = str((item or {}).get("path") or "")
        content = str((item or {}).get("content") or "")
        validation = validate_canonical_harper_artifact(path, content, evidence_text=evidence_text)
        if validation and not validation.get("ok"):
            rejected.append(validation)
            rejected_source_files.append(item)
            continue
        accepted.append(item)
    return {
        "accepted_files": accepted,
        "rejected": rejected,
        "rejected_source_files": rejected_source_files,
    }


def safe_rejected_artifact_name(path: str) -> str:
    normalized = str(path or "").replace("\\", "/").lstrip("./").lstrip("/")
    safe = re.sub(r"[^A-Za-z0-9._-]+", "__", normalized)
    safe = re.sub(r"[.]{2,}", "_", safe).strip("_")
    return (safe[:180] or "artifact") + ".invalid.md"


def rejected_artifact_debug_path(
    *,
    telemetry_root: str | Path,
    project_id: str,
    run_id: str | None,
    phase: str,
    artifact_path: str,
) -> Path:
    def safe_part(value: str, fallback: str) -> str:
        cleaned = re.sub(r"[^A-Za-z0-9._-]+", "_", str(value or fallback))
        cleaned = re.sub(r"[.]{2,}", "_", cleaned).strip("._")
        return cleaned or fallback

    root = Path(telemetry_root).resolve()
    candidate = (
        root
        / "rejected"
        / safe_part(project_id, "default")
        / safe_part(run_id or "n-a", "n-a")
        / safe_part(phase, "phase")
        / safe_rejected_artifact_name(artifact_path)
    )
    resolved = candidate.resolve()
    if root not in resolved.parents and resolved != root:
        raise ValueError("Rejected artifact debug path escaped telemetry root.")
    return resolved


def attach_rejected_artifact_debug_refs(
    rejected: List[Dict[str, Any]],
    *,
    telemetry_root: str | Path,
    project_id: str,
    run_id: str | None,
    phase: str,
    files: List[Dict[str, Any]],
) -> List[Dict[str, Any]]:
    by_path = {
        str((item or {}).get("path") or ""): str((item or {}).get("content") or "")
        for item in files or []
    }
    enriched: List[Dict[str, Any]] = []
    for item in rejected or []:
        cloned = dict(item)
        artifact_path = str(cloned.get("path") or "")
        content = by_path.get(artifact_path, "")
        try:
            debug_path = rejected_artifact_debug_path(
                telemetry_root=telemetry_root,
                project_id=project_id,
                run_id=run_id,
                phase=phase,
                artifact_path=artifact_path,
            )
            debug_path.parent.mkdir(parents=True, exist_ok=True)
            debug_path.write_text(content, encoding="utf-8")
            cloned["debug_path"] = debug_path.as_posix()
            cloned["rejected_artifact_ref"] = debug_path.as_posix()
        except Exception as exc:
            cloned["debug_path_error"] = str(exc)
        enriched.append(cloned)
    return enriched


def validate_current_canonical_core_blobs(
    core_blobs: Dict[str, str] | None,
) -> Dict[str, Any]:
    trusted: Dict[str, str] = {}
    invalid: List[Dict[str, Any]] = []
    for key, value in (core_blobs or {}).items():
        validation = validate_canonical_harper_artifact(str(key), str(value or ""))
        if validation and not validation.get("ok"):
            item = dict(validation)
            item["current_canonical_invalid"] = True
            item["untrusted_repair_material_snippet"] = str(value or "")[:2000]
            invalid.append(item)
            continue
        trusted[str(key)] = str(value or "")
    return {"trusted_core_blobs": trusted, "invalid_canonical": invalid}

File: extensions/vscode/harper-canonical-validation.js
const path = require('path');

const PROMPT_TEMPLATE_PHRASES = [
  'Print EXCLUSIVELY ' + 'one file block',
  'Produce only ' + 'the single',
  'single ' + 'BEGIN_FILE',
  'No additional ' + 'files',
];

const UNRESOLVED_PLACEHOLDERS = [
  '<Project Name>',
  '<X min',
  '<...>',
  'My Solution Name',
];

const GENERIC_TEMPLATE_ONLY_VALUES = [
  'aws-eks',
  'my-project-key',
  'https://api.openai.com',
];

function result(filePath, failedChecks, diagnostic) {
  return {
    ok: failedChecks.length === 0,
    path: filePath,
    failed_checks: failedChecks,
    diagnostic: failedChecks.length ? diagnostic : '',
    error_code: failedChecks.length ? 'invalid_canonical_artifact' : null,
  };
}

function normalizeOutputPath(filePath) {
  return String(filePath || '').replace(/\\/g, '/').replace(/^\.?\//, '').replace(/^\/+/, '');
}

function hasHeading(text, heading) {
  const escaped = heading.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  return new RegExp(`^\\s*${escaped}\\s*$`, 'im').test(text);
}

function headingIndex(text, heading) {
  const escaped = heading.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const match = new RegExp(`^\\s*${escaped}\\s*$`, 'im').exec(text);
  return match ? match.index : -1;
}

function sectionAfter(text, heading) {
  const escaped = heading.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const match = new RegExp(`^\\s*${escaped}\\s*$`, 'im').exec(text);
  if (!match) return '';
  const rest = text.slice(match.index + match[0].length);
  const next = /^\s*##\s+/m.exec(rest);
  return next ? rest.slice(0, next.index) : rest;
}

function commonMarkdownFailures(content, evidenceText = '') {
  const text = String(content || '');
  const lower = text.toLowerCase();
  const failures = [];
  if (text.includes('BEGIN_FILE')) failures.push('contains_BEGIN_FILE');
  if (text.includes('END_FILE')) failures.push('contains_END_FILE');
  for (const phrase of PROMPT_TEMPLATE_PHRASES) {
    if (lower.includes(phrase.toLowerCase())) {
      failures.push(`contains_prompt_template_phrase:${phrase}`);
    }
  }
  for (const placeholder of UNRESOLVED_PLACEHOLDERS) {
    if (text.includes(placeholder)) {
      failures.push(`contains_unresolved_placeholder:${placeholder}`);
    }
  }
  for (const value of GENERIC_TEMPLATE_ONLY_VALUES) {
    if (text.includes(value) && !String(evidenceText || '').includes(value)) {
      failures.push(`contains_unevidenced_template_value:${value}`);
    }
  }
  return failures;
}

function validateIdeaMarkdown(content, opts = {}) {
  const filePath = opts.path || 'docs/harper/IDEA.md';
  const evidenceText = opts.evidenceText || '';
  const text = String(content || '');
  const stripped = text.trimStart();
  const failures = commonMarkdownFailures(text, evidenceText);

  if (!(stripped.startsWith('# IDEA — ') || stripped.startsWith('# IDEA - '))) {
    failures.push('missing_idea_h1');
  }
  if (stripped.startsWith('```') || stripped.toLowerCase().startsWith('tech_constraints:')) {
    failures.push('starts_with_raw_yaml_or_fence');
  }

  const primary = [
    ['## Vision'],
    ['## Problem Statement'],
    ['## Target Users & Context'],
    ['## Value & Outcomes', '## Value & Outcomes (with initial targets)'],
    ['## Out of Scope', '## Out of Scope (slice-1)'],
    ['## Technology Constraints', '## Technology Constraints (SPEC-ready)'],
    ['## Risks & Assumptions'],
  ];
  const indexes = [];
  for (const options of primary) {
    const found = options.map((heading) => headingIndex(text, heading)).filter((idx) => idx >= 0);
    if (!found.length) {
      failures.push(`missing_heading:${options.join('|')}`);
    } else {
      indexes.push(Math.min(...found));
    }
  }
  if (indexes.length && indexes.some((idx, pos) => pos > 0 && idx < indexes[pos - 1])) {
    failures.push('primary_idea_sections_out_of_order');
  }
  const techSection = sectionAfter(text, '## Technology Constraints') || sectionAfter(text, '## Technology Constraints (SPEC-ready)');
  if (!/^```ya?ml\s*$/im.test(techSection)) {
    failures.push('missing_fenced_yaml_after_technology_constraints');
  }
  if (!/^\s*##\s+.*success metrics/im.test(text)) {
    failures.push('missing_success_metrics_section');
  }
  const coreOpeningIndexes = ['## Vision', '## Problem Statement', '## Target Users & Context']
    .map((heading) => headingIndex(text, heading))
    .filter((idx) => idx >= 0);
  const lastCoreOpening = coreOpeningIndexes.length ? Math.max(...coreOpeningIndexes) : -1;
  for (const heading of [
    '## Deployment Portability Rule',
    '## Technology Constraints Profile Rule',
    '## Strategic Fit',
    '## /spec Handoff Readiness',
  ]) {
    const idx = headingIndex(text, heading);
    if (idx >= 0 && lastCoreOpening >= 0 && idx < lastCoreOpening) {
      failures.push(`bmad_extra_section_before_primary:${heading}`);
    }
  }
  return result(filePath, failures, 'IDEA.md failed canonical Harper structure validation.');
}

function validateSpecMarkdown(content, opts = {}) {
  const filePath = opts.path || 'docs/harper/SPEC.md';
  const text = String(content || '');
  const lower = text.toLowerCase();
  const failures = commonMarkdownFailures(text);
  if (!text.trimStart().startsWith('# SPEC')) failures.push('missing_spec_h1');
  if (!['problem', 'scope', 'objective', 'requirement'].some((term) => lower.includes(term))) {
    failures.push('missing_problem_scope_or_requirements');
  }
  if (!['acceptance criteria', 'testability', 'test strategy'].some((term) => lower.includes(term))) {
    failures.push('missing_acceptance_or_testability');
  }
  if (!['constraint', 'non-functional', 'non functional'].some((term) => lower.includes(term))) {
    failures.push('missing_constraints_or_non_functional_requirements');
  }
  if (lower.includes('spec_ux_appendix') || (lower.includes('user journey') && !lower.includes('functional requirement'))) {
    failures.push('looks_like_companion_only_ux_content');
  }
  return result(filePath, failures, 'SPEC.md failed canonical Harper structure validation.');
}

function validatePlanMarkdown(content, opts = {}) {
  const filePath = opts.path || 'docs/harper/PLAN.md';
  const text = String(content || '');
  const lower = text.toLowerCase();
  const failures = commonMarkdownFailures(text);
  if (!text.trimStart().startsWith('# PLAN')) failures.push('missing_plan_h1');
  if (!text.includes('REQ-')) failures.push('missing_req_ids');
  if (!['dependencies', 'dependson', 'ordering', 'depends on'].some((term) => lower.includes(term))) {
    failures.push('missing_dependencies_or_ordering');
  }
  if (!['/kit', 'kit readiness', 'kit-readiness', 'implementation readiness'].some((term) => lower.includes(term))) {
    failures.push('missing_kit_or_implementation_readiness');
  }
  return result(filePath, failures, 'PLAN.md failed canonical Harper structure validation.');
}

function validatePlanJson(content, opts = {}) {
  const filePath = opts.path || 'docs/harper/plan.json';
  const text = String(content || '');
  const failures = [];
  if (text.trimStart().startsWith('#') || text.includes('```')) failures.push('looks_like_markdown');
  let data;
  try {
    data = JSON.parse(text);
  } catch {
    failures.push('invalid_json');
    return result(filePath, failures, 'plan.json must be valid JSON.');
  }
  const reqs = data && (data.reqs || data.requirements || data.items);
  if (!Array.isArray(reqs) || !reqs.length) {
    failures.push('missing_requirements_list');
  } else {
    reqs.forEach((req, index) => {
      if (!req || typeof req !== 'object' || Array.isArray(req)) {
        failures.push(`req_${index}_not_object`);
        return;
      }
      for (const field of ['id', 'title', 'status', 'acceptance']) {
        if (!req[field]) failures.push(`req_${index}_missing_${field}`);
      }
      if (!('dependsOn' in req) && !('dependencies' in req) && !('depends_on' in req)) {
        failures.push(`req_${index}_missing_dependencies`);
      }
    });
  }
  return result(filePath, failures, 'plan.json failed canonical Harper structure validation.');
}

function validateLaneGuideMarkdown(content, opts = {}) {
  const filePath = opts.path || 'docs/harper/lane-guides/<lane>.md';
  const text = String(content || '');
  const lower = text.toLowerCase();
  const failures = commonMarkdownFailures(text);
  if (!text.trimStart().startsWith('#')) failures.push('missing_heading');
  if (!['lane purpose', 'purpose', 'scope'].some((term) => lower.includes(term))) {
    failures.push('missing_lane_purpose_or_scope');
  }
  if (!['expected files', 'boundaries', 'boundary'].some((term) => lower.includes(term))) {
    failures.push('missing_expected_files_or_boundaries');
  }
  if (!['test command', 'validation command', 'commands'].some((term) => lower.includes(term))) {
    failures.push('missing_test_or_validation_commands');
  }
  if (!['eval/gate', 'eval expectations', 'gate expectations'].some((term) => lower.includes(term))) {
    failures.push('missing_eval_gate_expectations');
  }
  return result(filePath, failures, 'Lane guide failed canonical Harper structure validation.');
}

function validateCanonicalHarperArtifact(filePath, content, opts = {}) {
  const normalized = normalizeOutputPath(filePath);
  if (normalized === 'docs/harper/IDEA.md') return validateIdeaMarkdown(content, { ...opts, path: normalized });
  if (normalized === 'docs/harper/SPEC.md') return validateSpecMarkdown(content, { ...opts, path: normalized });
  if (normalized === 'docs/harper/PLAN.md') return validatePlanMarkdown(content, { ...opts, path: normalized });
  if (normalized === 'docs/harper/plan.json') return validatePlanJson(content, { ...opts, path: normalized });
  if (/^docs\/harper\/lane-guides\/[^/]+\.md$/i.test(normalized)) {
    return validateLaneGuideMarkdown(content, { ...opts, path: normalized });
  }
  return null;
}

function safeRejectedFileName(filePath) {
  return normalizeOutputPath(filePath)
    .replace(/[^A-Za-z0-9._-]+/g, '__')
    .replace(/[.]{2,}/g, '_')
    .replace(/^_+/, '')
    .slice(0, 180) || 'artifact';
}

function rejectedHarperArtifactPath({ phase, runId, filePath }) {
  const phasePart = String(phase || 'unknown').replace(/[^A-Za-z0-9._-]+/g, '_').replace(/[.]{2,}/g, '_');
  const runPart = String(runId || 'unknown').replace(/[^A-Za-z0-9._-]+/g, '_').replace(/[.]{2,}/g, '_');
  return path.posix.join(
    '.clike',
    'rejected',
    'harper',
    phasePart,
    runPart,
    `${safeRejectedFileName(filePath)}.invalid.md`
  );
}

function stringifyHarperValue(value) {
  if (value == null) return '';
  if (typeof value === 'string') return value;
  if (value instanceof Error) return value.message || String(value);
  if (Array.isArray(value)) {
    return value.map((item) => stringifyHarperValue(item)).filter(Boolean).join('\n');
  }
  if (typeof value === 'object') {
    if (typeof value.detail === 'string') return value.detail;
    if (value.detail && typeof value.detail === 'object') return stringifyHarperValue(value.detail);
    if (typeof value.message === 'string') return value.message;
    if (typeof value.text === 'string') return value.text;
    try {
      return JSON.stringify(value, null, 2);
    } catch {
      return Object.prototype.toString.call(value);
    }
  }
  return String(value);
}

function unwrapHarperFailure(value) {
  if (value && typeof value === 'object' && !(value instanceof Error)) {
    if (value.out && typeof value.out === 'object') return value.out;
    if (value.detail && typeof value.detail === 'object') return value.detail;
  }
  return value;
}

function normalizeHarperFailure(errorOrResult) {
  const result = unwrapHarperFailure(errorOrResult);
  if (typeof result === 'string' || result instanceof Error || !result || typeof result !== 'object') {
    return {
      ok: false,
      phase: undefined,
      error_code: undefined,
      text: stringifyHarperValue(result),
      errors: [],
      warnings: [],
      rejected: [],
      partial_files: [],
      diagnostic_files: [],
    };
  }

  const errors = Array.isArray(result.errors) ? result.errors : [];
  const warnings = Array.isArray(result.warnings) ? result.warnings : [];
  const rejected = Array.isArray(result.rejected) ? result.rejected : [];
  const partialFiles = Array.isArray(result.partial_files)
    ? result.partial_files
    : (Array.isArray(result.diagnostic_files) ? result.diagnostic_files : []);

  return {
    ok: result.ok === false ? false : Boolean(result.error_code || errors.length || rejected.length),
    phase: result.phase,
    error_code: result.error_code,
    text: typeof result.text === 'string' ? result.text : stringifyHarperValue(result),
    errors,
    warnings,
    rejected,
    partial_files: partialFiles,
    diagnostic_files: Array.isArray(result.diagnostic_files) ? result.diagnostic_files : partialFiles,
    runId: result.runId || result.run_id,
  };
}

function formatHarperError(errorOrResult) {
  const failure = normalizeHarperFailure(errorOrResult);
  const phase = failure.phase ? ` ${failure.phase}` : '';
  const code = failure.error_code || 'error';
  const lines = [`Harper${phase} failed: ${code}`];

  if (failure.text) {
    lines.push('', failure.text);
  }

  const diagnostics = [];
  const diagnosticItems = [
    ...failure.rejected,
    ...failure.errors.filter((item) => item && typeof item === 'object'),
  ];
  for (const item of diagnosticItems) {
    for (const check of (Array.isArray(item.failed_checks) ? item.failed_checks : [])) {
      diagnostics.push(String(check));
    }
  }
  if (diagnostics.length) {
    lines.push('', 'Failed checks:');
    for (const check of [...new Set(diagnostics)]) {
      lines.push(`- ${check}`);
    }
  }

  if (failure.rejected.length) {
    lines.push('', 'Rejected:');
    for (const item of failure.rejected) {
      const filePath = item && item.path ? item.path : stringifyHarperValue(item);
      lines.push(`- ${filePath}`);
      const debugPath = item && (item.debug_path || item.rejected_artifact_ref);
      if (debugPath) lines.push(`  debug: ${debugPath}`);
    }
  }

  const stringErrors = failure.errors
    .filter((item) => typeof item === 'string' && item.trim())
    .map((item) => item.trim());
  if (stringErrors.length) {
    lines.push('', 'Errors:');
    for (const item of stringErrors) lines.push(`- ${item}`);
  }

  if (failure.partial_files.length) {
    lines.push('', 'Partial files not applied:');
    for (const item of failure.partial_files.slice(0, 20)) {
      lines.push(`- ${item.path || stringifyHarperValue(item)}`);
    }
  }

  return lines.join('\n');
}

function formatInvalidCanonicalArtifactMessage(out) {
  const rejected = Array.isArray(out && out.rejected) ? out.rejected : [];
  const errors = Array.isArray(out && out.errors) ? out.errors : [];
  const items = rejected.length
    ? rejected
    : errors.filter((item) => item && typeof item === 'object' && item.error_code === 'invalid_canonical_artifact');

  const lines = [
    'CLike blocked a generated canonical Harper artifact because it failed structural validation.',
    'No canonical Harper file was overwritten.',
  ];

  if (items.length) {
    for (const item of items.slice(0, 5)) {
      const filePath = item.path || 'canonical artifact';
      const checks = Array.isArray(item.failed_checks) ? item.failed_checks.join(', ') : 'unknown';
      lines.push(`- ${filePath}: ${checks}`);
      const debugPath = item.debug_path || item.rejected_artifact_ref;
      if (debugPath) {
        lines.push(`  rejected debug path: ${debugPath}`);
      }
    }
  } else if (out && out.text) {
    lines.push(String(out.text));
  }

  lines.push('Repair the canonical artifact shape and rerun the Harper phase.');
  return lines.join('\n');
}

module.exports = {
  validateIdeaMarkdown,
  validateSpecMarkdown,
  validatePlanMarkdown,
  validatePlanJson,
  validateLaneGuideMarkdown,
  validateCanonicalHarperArtifact,
  rejectedHarperArtifactPath,
  stringifyHarperValue,
  normalizeHarperFailure,
  formatHarperError,
  formatInvalidCanonicalArtifactMessage,
};
